docs/dev/testing.rst

```diff
@@ -0,0 +1,306 @@
+.. _testing:
+
+===============================================================================
+Testing PUDL
+===============================================================================
+Our test suite is primarily written using `pytest <https://pytest.org>`__.
+The tests are all stored under the ``test/`` directory in the main repository,
+and they are organized into 3 broad categories, each with its own subdirectory:
+
+* **Software Unit Tests** (``test/unit/``) can be run in seconds and don't
+  require any external data. They test the basic functionality of various
+  functions and classes, often using minimal inline data structures that are
+  specified in the test modules themselves.
+* **Software Integration Tests** (``test/integration/``) test larger
+  collections of functionality, including the interactions between different
+  parts of the overall software system, and in some cases interactions with
+  external systems, requiring network connectivity. The main thing our
+  integration tests do is run the full PUDL data processing pipeline for the
+  most recent year of data. This takes around 15 minutes.
+* **Data Validations** (``test/validate/``) sanity check the PUDL outputs
+  generated by the data processing pipeline. This helps us catch issues with
+  the input data, and more subtle bugs that don't prevent the code from
+  executing, but do have unintended or unexpected impacts on the output data.
+  The data validation requires a fully populated PUDL database and is quite
+  different from the other tests.
+
+-------------------------------------------------------------------------------
+Running tests with Tox
+-------------------------------------------------------------------------------
+We use `Tox <https://tox.readthedocs.io>`__ to coordinate running different
+groups of tests together, and to make sure that the tests are run in a fresh
+Python environment without any other packages that might have been installed
+locally, but that won't necessarily be installed by PUDL users. Tox's behavior
+is configured with the ``tox.ini`` file in the main repository directory. There
+are several different "test environments" defined, to test different aspects
+of the software, or to perform other actions like building the documentation.
+To see a list of the available environments with short descriptions in the PUDL
+repository run:
+
+.. code-block:: console
+
+    $ tox -av
+
+    default environments:
+    ci               -> Run all continuous integration (CI) checks & generate test coverage.
+
+    additional environments:
```

Comment:
This is literally the output from that command, cut-and-pasted into this code-block. I don't have any control over the formatting and am trying to avoid documenting the full complexity of all of the environments we've set up in there beyond the 1-line descriptions which are embedded in the file, and available through this command.

Reply:
ah. okay. this makes sense, but is a little confusing.  I would say it is fine given the efficiency but not ideal.